File: app/src/main/java/edu/gatech/cog/MainActivity.java
package edu.gatech.cog;

import android.app.Activity;
import android.content.Context;
import android.media.AudioManager;
import android.os.Bundle;
import android.view.View;
import android.view.ViewGroup;
import android.widget.AdapterView;

import com.google.android.glass.media.Sounds;
import com.google.android.glass.widget.CardBuilder;
import com.google.android.glass.widget.CardScrollAdapter;
import com.google.android.glass.widget.CardScrollView;

public class MainActivity extends Activity {

    private CardScrollView cardScroller;

    private View view;

    @Override
    protected void onCreate(Bundle bundle) {
        super.onCreate(bundle);

        view = buildView();

        cardScroller = new CardScrollView(this);
        cardScroller.setAdapter(new CardScrollAdapter() {
            @Override
            public int getCount() {
                return 1;
            }

            @Override
            public Object getItem(int position) {
                return view;
            }

            @Override
            public View getView(int position, View convertView, ViewGroup parent) {
                return view;
            }

            @Override
            public int getPosition(Object item) {
                if (view.equals(item)) {
                    return 0;
                }
                return AdapterView.INVALID_POSITION;
            }
        });

        cardScroller.setOnItemClickListener(new AdapterView.OnItemClickListener() {
            @Override
            public void onItemClick(AdapterView<?> parent, View view, int position, long id) {
                AudioManager am = (AudioManager) getSystemService(Context.AUDIO_SERVICE);
                am.playSoundEffect(Sounds.DISALLOWED);
            }
        });
        setContentView(cardScroller);
    }

    @Override

    protected void onResume() {
        super.onResume();
        cardScroller.activate();
    }

    @Override
    protected void onPause() {
        cardScroller.deactivate();
        super.onPause();
    }

    private View buildView() {
        CardBuilder card = new CardBuilder(this, CardBuilder.Layout.TEXT);

        card.setText(R.string.hello_world);
        return card.getView();
    }

}

File: app/src/main/java/edu/gatech/cog/MainActivity.kt
package edu.gatech.cog

import android.Manifest
import android.content.pm.PackageManager
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.WindowManager
import androidx.core.app.ActivityCompat
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.LinearLayoutManager
import com.microsoft.cognitiveservices.speech.SpeechConfig
import com.microsoft.cognitiveservices.speech.SpeechRecognizer
import kotlinx.android.synthetic.main.activity_main.*
import kotlin.concurrent.thread

class MainActivity : AppCompatActivity() {

    private val TAG = "MainActivity"
    private val speechSubscriptionKey = ""
    private val serviceRegion = ""
    private var speechRecognizer: SpeechRecognizer? = null

    private lateinit var adapter: TextAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        if (ContextCompat.checkSelfPermission(
                this,
                Manifest.permission.RECORD_AUDIO
            ) != PackageManager.PERMISSION_GRANTED
        ) {
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.RECORD_AUDIO), 1)
        }

        window.addFlags(WindowManager.LayoutParams.FLAG_KEEP_SCREEN_ON)

        val content = mutableListOf<String>()
        adapter = TextAdapter(content)
        rvText.adapter = adapter
        val layoutManager = LinearLayoutManager(this)
        layoutManager.stackFromEnd = true
        rvText.layoutManager = layoutManager

        thread {
            val config = SpeechConfig.fromSubscription(speechSubscriptionKey, serviceRegion)
            speechRecognizer = SpeechRecognizer(config)
            speechRecognizer?.startContinuousRecognitionAsync()
            speechRecognizer?.recognizing?.addEventListener { _, speechRecognitionEventArgs ->
                Log.v(TAG, "Recognizing: ${speechRecognitionEventArgs.result.text}")
                runOnUiThread {
                    tvText.text = speechRecognitionEventArgs.result.text
                }
            }
            speechRecognizer?.recognized?.addEventListener { _, speechRecognitionEventArgs ->
                Log.v(TAG, "Recognized: ${speechRecognitionEventArgs.result.text}")
                content.add(speechRecognitionEventArgs.result.text)
                runOnUiThread {
                    tvText.text = ""
                    adapter.notifyItemInserted(content.size)
                    rvText.scrollToPosition(content.size - 1)

                    Log.v(TAG, "Transcript: $content")
                }
            }
        }
    }

    override fun onResume() {
        super.onResume()
        speechRecognizer?.startContinuousRecognitionAsync()
    }

    override fun onPause() {
        super.onPause()
        speechRecognizer?.close()
    }
}
